### ISSUE TITLE
Memory.Initialization: Deprecate moveInitializeArrayElements

### ISSUE THREAD

**Issue by benharsh** [chapel-lang/chapel#20339]
The current functions named the ``moveArrayElements`` are:

```chpl
// move within a single array
proc moveInitializeArrayElements(ref a: [?d], dstStartIndex: a.idxType,
                                 srcStartIndex: a.idxType,
                                 numElements: int)

// move between two arrays
proc moveInitializeArrayElements(ref dstA: [] ?t,
                                 dstStartIndex: dstA.idxType,
                                 srcA: [] t,
                                 srcStartIndex: srcA.idxType,
                                 numElements: int)
```

These functions halt, and rather than accepting ranges or domains they accept starting indices and a number of elements. I do not believe we would wish to support this kind of procedure for a 2.0 library.

I propose deprecating these functions. A replacement is proposed in https://github.com/chapel-lang/chapel/issues/20338. I expect this replacement to at the very least be an "unstable" member of the Memory.Initialization module.

**Comment by mppf** [comment]
I'm happy with the direction of deprecating these and creating unstable replacements for them as described in #20338.

**Comment by benharsh** [comment]
> I'm happy with the direction of deprecating these and creating unstable replacements for them as described in #20338.

On that issue you mentioned needing/wanting that feature for some sort code. Could you use that in Sort if it was marked unstable in a way that resulted in a compiler warning? I was assuming that's what unstable meant in this context.

**Comment by mppf** [comment]
I would expect that code we maintain does not need to produce warnings when using unstable features with the regular `--warn-unstable` but that another flag could enable that. But, we can work around that with other no-doc routines etc, if it does not work that way.

**Comment by lydia-duncan** [comment]
To clarify, are you saying that we'd individually turn off unstable warnings for internal/standard/package module uses of unstable symbols that we're okay with keeping for a little longer?  Or that you'd expect a flag that squashes all unstable warnings in internal/standard/package modules?  I'm more okay with the former, though I still think using it would bite us

**Comment by mppf** [comment]
I was imagining the default would be to not warn for uses of unstable features in internal/standard/package modules (since it is code we maintain & so there is no stability concern -- we will keep it working) and for a flag to activate that (so we can check). However I don't have any strong opinion about this and there are many other ways to handle it.

**Comment by bradcray** [comment]
Another approach would be to created an undocumented + stable version of the routine that Sort calls, and to have the current public/user-facing routine generate the warning and call the undocumented version.

Michael's proposal is intriguing to me, though I think we'd also want the ability to check our own modules for uses of unstable features.  But maybe that's just a question of having `--devel` + `--warn-unstable` do so (?).  (though that might be hard to run through parallel testing?)

**Comment by mppf** [comment]
> But maybe that's just a question of having --devel + --warn-unstable do so (?). (though that might be hard to run through parallel testing?

Yeah, IMO it'd be OK for `--devel` to set another flag, but I think we would need an additional flag for it so that we can test without worrying about differences in output from `--devel`. The other flag could be e.g. `--warn-unstable-builtin-modules` and I would want it to be off by default.

**Comment by benharsh** [comment]
Following subteam discussion, we have reached consensus on deprecating these two functions in favor of the proposal in https://github.com/chapel-lang/chapel/issues/20338. It was felt that the signature of these functions is not something we would want to support in the long-term, and that the new proposal was a satisfactory improvement provided it remain unstable until we have more time to clarify certain design questions.

**Comment by benharsh** [comment]
This has been resolved by #21758.
